addon/v1/types.go

```diff
@@ -0,0 +1,178 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedClusterAddon is the Custom Resource object which holds the current state
+// of an operator. This object is used by operators to convey their state to
+// the rest of the cluster.
+type ManagedClusterAddon struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// spec holds configuration that could apply to any operator.
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ManagedClusterAddonSpec `json:"spec"`
+
+	// status holds the information about the state of an operator.  It is consistent with status information across
+	// the Kubernetes ecosystem.
+	// +optional
+	Status ManagedClusterAddonStatus `json:"status"`
+}
+
+// ManagedClusterAddonSpec is empty for now, but you could imagine holding information like "pause".
+type ManagedClusterAddonSpec struct {
+	// UpdateApproved ...
+	UpdateApproved bool `json:"updateApproved"`
+}
+
+// ManagedClusterAddonStatus provides information about the status of the operator.
+// +k8s:deepcopy-gen=true
+type ManagedClusterAddonStatus struct {
+	// conditions describes the state of the operator's managed and monitored components.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +optional
+	Conditions []AddonStatusCondition `json:"conditions,omitempty"  patchStrategy:"merge" patchMergeKey:"type"`
+
+	// relatedObjects is a reference to the configuration resource that drives the operator
+	// +optional
+	RelatedObject ObjectReference `json:"relatedObject,omitempty"`
+
+	// LatestRelease indicates the latest available release for the addon
+	LatestRelease Release `json:"availableUpdate,omitempty"`
+
+	// CurrentRelease ...
```

Comment:
you should provide guidance about how to handle "what version is this" during an upgrade when multiple versions are involved.  The rule changed recently in openshift to be, "update the currentVersion **after** you are fully rolled out"

Reply:
added detail description